qutip/core/qobj.py

```diff
@@ -296,58 +271,110 @@ class Qobj:
         else:
             self._data = _data.create(arg, copy=copy)
+            if (
+                dims
+                and self._data.shape != dims.shape
+                and self._data.shape == dims.shape[::-1]
+            ):
+                self._data = _data.transpose(self._data)
             self.dims = dims or [[self._data.shape[0]], [self._data.shape[1]]]
 
     def __init__(self, arg=None, dims=None, type=None,
                  copy=True, superrep=None, isherm=None, isunitary=None):
-        self._initialize_data(arg, dims, copy)
-        self.type = type or type_from_dims(self.dims)
         self._isherm = isherm
         self._isunitary = isunitary
+        self._superrep = None
+        if isinstance(dims, list):
+            dims = Dimensions(dims)
+        self._initialize_data(arg, dims, copy)
 
-        if self.type == 'super' and type_from_dims(self.dims) == 'oper':
-            if self._data.shape[0] != self._data.shape[1]:
-                raise ValueError("".join([
-                    "cannot build superoperator from nonsquare data of shape ",
-                    repr(self._data.shape),
-                ]))
-            root = int(np.sqrt(self._data.shape[0]))
-            if root * root != self._data.shape[0]:
-                raise ValueError("".join([
-                    "cannot build superoperator from nonsquare subspaces ",
-                    "of size ",
-                    repr(self._data.shape[0]),
-                ]))
-            self.dims = [[[root]]*2]*2
-        if self.type in ['super', 'operator-ket', 'operator-bra']:
-            superrep = superrep or 'super'
+        # Dims are guessed from the data and need to be changed to super.
+        if (
+            type in ['super', 'operator-ket', 'operator-bra']
+            and self._dims.type in ['oper', 'ket', 'bra']
+        ):
+            root_right = int(np.sqrt(self._data.shape[0]))
+            root_left = int(np.sqrt(self._data.shape[1]))
+            if (
+                root_right * root_right != self._data.shape[0]
+                and root_left * root_left != self._data.shape[1]
+            ):
+                raise ValueError(
+                    "cannot build superoperator from nonsquare subspaces"
+                )
+            self.dims = [[[root_right]]*2, [[root_left]]*2]
+
+        self.type = type
         self.superrep = superrep
 
     def copy(self):
         """Create identical copy"""
         return Qobj(arg=self._data,
                     dims=self.dims,
-                    type=self.type,
-                    superrep=self.superrep,
                     isherm=self._isherm,
                     isunitary=self._isunitary,
                     copy=True)
 
+    @property
+    def dims(self):
+        return self._dims.as_list()
+
+    @dims.setter
+    def dims(self, dims):
+        dims = Dimensions(dims, rep=self.superrep)
+        if self._data and dims.shape != self._data.shape:
```

Comment:
I'd prefer it if `_dims` and `_data` were both always already set when the property getters and setters are called. Otherwise we support strange logic for partially initialized Qobj's throughout the methods.